docs/architecture/adr-042-blockchain-riri-org.md

```diff
@@ -0,0 +1,292 @@
+The Processor will maintain an internal cursor `height` of the last
+processed block. As a set of unordered blocks arrive, the the Processor
+will check if it has `height+1` nessary to process the next block.
+
+```go
+type Proccesor struct {
+    height ...
+    state ...
+    blocks [height]*Block
+    peers[height]PeerID
+    lastTouch timestamp
+}
+
+func (proc *Processor) handleBlockResponse(peerID, block, time) {
+    lastTouch = time
+    if block.height < height {
+        // skip
+    } else if blocks[block.height] {
+        return errDuplicateBlock{}
+    } else  {
+        blocks[block.height] = block
+    }
+
+    if blocks[height] && blocks[height+1] {
+        ... = processBlock(blocks[height], blocks[height+1])
+         bcR.store.SaveBlock(first, firstParts, second.LastCommit)
+        ... = proc.state.Validators.VerifyCommit
+        delete blocks[height]
+        height++
+    }
+    return pcBlockProcessed{height}
+}
+
+func (proc *Processor) handleRemovePeer(peerID) {
+    events = []
+    // Delete all unprocessed blocks from peerID
+    for i = height; i < len(blocks); i++ {
+        if peers[i] == peerID {
+            events = append(events, pcBlockReschedule{height})
+            delete block[height]
+        }
+    }
+    return events
+}
+
+func handleTimeCheckEv(time) {
+    if time - lastTouch > timeout {
+        // Timeout the processor
+        ...
+    }
+}
+```
+
+## Schedule
+The scheduler (previously the pool) is responsible for squeding peer status requests and block request responses.
+
+```go
+type schedule {
+    ...
+}
+
+type Scheduler struct {
+    ...
+    schedule schedule{}
+}
+
+func addPeer(peerID) {
+	schedule.addPeer(peerID)
+}
+
+func handleStatusResponse(peerID, height, time) {
+	schedule.touchPeer(peerID, time)
+	schedule.setPeerHeight(peerID, height)
+}
+
+func handleBlockResponseMessage(peerID, height, block, time) {
+	schedule.touchPeer(peerID, time)
+	schedule.markReceived(peerID, height, size(block))
+}
+
+func handleNoBlockResponseMessage(peerID, height, time) {
+	schedule.touchPeer(peerID, time)
+    // reschedule that block, punish peer...
+    ...
+}
+
+func handleTimeCheckEv(time) {
+	// clean peer list
+
+    events = []
+	for peerID := range schedule.peersTouchedSince(time) {
+		pending = schedule.pendingFrom(peerID) 
+		schedule.setPeerState(peerID, timedout)
+		schedule.resetBlocks(pending)
+		events = append(events, peerTimeout{peerID})
+    }
+
+	events = append(events, schedule.getSchedule())
+
+    return events
+}
+```
+
+## Peer
+The Peer Stores per peer state based on messages received by the scheduler.
+
+```go
+type Peer struct {
+	lastTouched timestamp
+	lastDownloaded timestamp
+	pending map[height]struct{}
+	height height // max height for the peer
+	state {
+		pending,   // we know the peer but not the height
+		active,    // we know the height
+		timeout    // the peer has timed out
+	}
+}
+```
+
```

Comment:
Yes I think we the `schedule.peersNotTouchedSince` loop could be done on `schedule.peersSlowerThan(minSpeed)` to prune peers in a more sophisticated way. It could be as simple as that  or if we may eventually need something more sophisticated. We could have a sort of `PeerPruning` strategy which could be applied to the schedule. Not sure if that's needed now or if we need to emit events indicating why the peer was pruned 🤔.

Reply:
Both v0 and v1 have a monitor and remove slow peers. Events are sent to the switch to remove the slow peer. It is probably ok if we want to leave this aside initially for this refactor.

Reply:
👍 Yeah I think we can do this without much trouble.